draco/analysis/transform.py

```diff
@@ -1087,31 +1106,206 @@ def process(self, polcont):
+                        out_dset[oslc] += dset[islc]
+
+                # Normalize the output based on how many polarisations were summed
+                if name == weight_dset_name:
+                    out_dset[oslc] = (
+                        flag * nsum**2 * tools.invert_no_zero(out_dset[oslc])
+                    )
+
+                elif np.issubdtype(out_dset.dtype, np.integer):
+                    out_dset[oslc] //= nsum
+
+                elif np.issubdtype(out_dset.dtype, np.bool_):
+                    pass
 
-                if np.issubdtype(out_dset.dtype, np.integer):
-                    out_dset[:] //= 2
                 else:
-                    out_dset[:] *= 0.5
+                    out_dset[oslc] /= nsum
+
+        return outcont
+
+
+class PolWeightedAverage(task.SingleTask):
+    """Compute an optimally weighted pseudo-Stokes I from XX and YY polarisations.
+
+    This computes a weighted average:
+        data_I = (w_XX * d_XX + w_YY * d_YY) / (w_XX + w_YY)
+        weight_I = w_XX + w_YY
+
+    Requires the input container to be a subclass of DataWeightContainer and to
+    contain both XX and YY polarisations.
+    """
+
+    def process(self, polcont):
+        """Compute pseudo-Stokes I from XX and YY.
+
+        Parameters
+        ----------
+        polcont : DataWeightContainer
+            A container with weight dataset and a 'pol' axis containing XX and YY.
+
+        Returns
+        -------
+        outcont : same type as polcont
+            A container with a single polarisation axis labeled 'I'.
+        """
+        # Check input container
+        if not hasattr(polcont, "_weight_dset_name"):
+            raise TypeError(
+                "Input must be a subclass of DataWeightContainer with defined weight datasets."
+            )
+
+        if "pol" not in polcont.axes:
+            raise ValueError(
+                f"Input container of type {type(polcont)} does not have a 'pol' axis."
+            )
+
+        input_pol = list(polcont.index_map["pol"])
+        if "XX" not in input_pol or "YY" not in input_pol:
+            raise ValueError("Input must contain both 'XX' and 'YY' polarisations.")
+
+        ixx = input_pol.index("XX")
+        iyy = input_pol.index("YY")
+
+        if iyy > ixx:
+            start = ixx
+            stride = iyy - ixx
+        else:
+            start = iyy
+            stride = ixx - iyy
+
+        pol_slice = slice(start, start + stride + 1, stride)
+
+        def make_pol_slice(axis_names):
+            axis = list(axis_names).index("pol")
+            slc = (slice(None),) * axis + (pol_slice,)
+            return axis, slc
+
+        # Create output container
+        outcont = containers.empty_like(polcont, pol=np.array(["I"]))
+
+        for name in polcont.datasets.keys():
+            if name not in outcont.datasets:
+                outcont.add_dataset(name)
+
+        outcont.redistribute("freq")
```

Comment:
should there be a corresponding call on `poldata` to ensure both are distributed along the same axis?

Reply:
Good call.  This is now fixed.